Adds tsc --init option

tinganho opened the pull request:
Fixes #3079.

Defaults to compiler options:
```
{
     "module": "commonjs",
     "target": "es3",
     "noImplicitAny": true,
     "rootDir": ".",
     "outDir": "built",
     "sourceMap": false,
}
```
The compiler options are overridable and you are able to define more options via the CLI(with the exception of the CLI options `help`, `watch`, `version`, `init`).

Defaults to no `files` property but if one invokes with files:
```
tsc --init test.ts
```
It will add a files property with the specified files.

Adds an exclude of `node_modules` folder only.

It outputs comments and trailing commas in `tsconfig.json`. So this PR also adds parsing/stripping of comments and trailing commas in `tsconfig.json` file.

tinganho commented:
I could alternatively put `tsconfig.json` emitter logic in a new file called `configEmitter.ts`? It is now in the `emitter.ts` which hosts the code for the main compiler emitter logic, which might not be so appropriate.

mhegazy commented on src/compiler/program.ts:
```diff
@@ -26,6 +26,12 @@ namespace ts {
         return undefined;
     }
 
+    export function writeConfigFile(file: string, compilerOptions: CompilerOptions, fileNames: string[]): void {
```
I think this should be in tsc.ts, i would even inline it there.

mhegazy commented on src/compiler/emitter.ts:
```diff
@@ -3,6 +3,190 @@
 /* @internal */
 namespace ts {
+    export const defaultInitCompilerOptions: CompilerOptions = {
+        module: ModuleKind.CommonJS,
```
default is none.

tinganho replied:
Don't you think it should be `commonjs` since it is the most common? I think most project is either commonjs or amd.

tinganho replied:
VSCode also autocompletes `commonjs` as the default.

mhegazy replied:
i think this is not correct. @jrieken can you comment.

mhegazy commented on src/compiler/emitter.ts:
```diff
@@ -3,6 +3,190 @@
+    export const defaultInitCompilerOptions: CompilerOptions = {
+        module: ModuleKind.CommonJS,
+        target: ScriptTarget.ES3,
+        noImplicitAny: true,
```
default is false

mhegazy commented on src/compiler/emitter.ts:
```diff
@@ -3,6 +3,190 @@
+        noImplicitAny: true,
+        outDir: "built",
+        rootDir: ".",
+        sourceMap: false,
```
why have this even in the defaults?

tinganho replied:
Because it should be easy to switch otherwise people might know if it even exists. In VSCode they also show this as the default.

mhegazy commented on src/compiler/emitter.ts:
```diff
@@ -3,6 +3,190 @@
+        sourceMap: false,
+    }
+
+    export function buildConfigFile(writer: EmitTextWriter, compilerOptions: CompilerOptions, fileNames?: string[], excludes?: string[]) {
```
I do not think this belongs in the emitter. i think this should be commandlineParser.ts

tinganho replied:
`commandLineParser.ts` host the code for parsing logic, so I don't think it fits in their either. What do you think about creating a new file `configEmitter.ts`? That's probably more inline with `declarationEmitter.ts`.

mhegazy commented on src/compiler/emitter.ts:
```diff
@@ -3,6 +3,190 @@
+        let { optionNameMap } = getOptionNameMap();
+        writeTsConfigDotJsonFile();
+
+        function writeTsConfigDotJsonFile() {
```
can we split this change into two, one that does the `--init` and one that does the comments in json

mhegazy replied:
I would just do JSON.stringify after replacing the enum vales with their names. i do not think the comments add that much value here.

tinganho replied:
I would prefer to have at least the comments. Because people might not know what options do what. This is more inline with VSCode they also support comments and they come as default in the debug settings explaining every option.

Trailing comma is not supported in VSCode though. I could probably remove it.

mhegazy commented on src/compiler/commandLineParser.ts:
```diff
@@ -346,12 +368,114 @@ namespace ts {
+      * Remove whitespace, comments and trailing commas from JSON text.
+      * @param text JSON text string.
+      */
+    function stripJsonTrivia(text: string): string {
```
I would split this out to a different change for now.

tinganho commented:
@mhegazy could you please reconsider supporting comments. In addition to giving people verbose description, people might want to make annotations themselves in the config file. I can rework my PR if it it doesn't fit TS.

mhegazy commented:
I am fine doing comments. i just want them in a different change, it will make it easier to focus on one issue at a time, i.e. what are the defaults or the right things to emit, instead of talking about comments and parsing.

In general i think writing the output should be simpler. for JSON comments, i would be incline to use the scanner to write a small JSON parser, instead of removing comments.

tinganho commented:
>In general i think writing the output should be simpler. for JSON comments, i would be incline to use the scanner to write a small JSON parser, instead of removing comments.

But then you would need to create a JSON parser? Right now, I think it is easier to remove the comments and pass them to `JSON.parse`.

tinganho commented:
I just updated the PR. 

 - Removed trailing commas. Because VSCode doesn't support it
 - Moved emitter logic to `program.ts`.
 - Removed trailing commas.
 - Adds a success message after `tsc --init`.

Can we skip splitting the comments in another PR? It is really a small addition(15-20 LOC) to this PR and splitting it out means quite much work for me. I mean saving it another branch and then to remember the branch and merge later.

tinganho commented:
ping @mhegazy

mhegazy commented on src/compiler/program.ts:
```diff
@@ -26,6 +27,221 @@ namespace ts {
+        let { optionNameMap } = getOptionNameMap();
+        writeConfigFile();
+
+        function writeConfigFile() {
```
i do not think we need a full JSON serializer. That was my inital comment. for me this function, without the support for comments, can be written as:
```ts


JSON.stringify(opt, (k, v) => {
    switch (k) {
        case "target": return scriptTargetToString(k);
        case "module": return moduleKindToString(k);
        case "init":
        case "watch":
        case "version":
        case "help":
            return undefined;
        default: return v;
    }
}, "    ");
```

mhegazy commented on src/compiler/tsc.ts:
```diff
@@ -144,6 +144,20 @@ namespace ts {
         let timerHandle: number;                    // Handle for 0.25s wait timer
 
+        if (commandLine.options.init) {
+            let file = `${sys.getCurrentDirectory()}/tsconfig.json`;
```
use `combinePaths` instead of adding the `/` explicitlly

mhegazy commented:
sorry for the delay. let me clarify my comments earlier..

1. Please split the `stripJsonTrivia` into a different PR. i do not think we need it to support --init. 
2. The JSON serialization can be much simpler if you do not write comments in the file. I believe this is the right thing to do as the file is not a valid JSON file at this point.

tinganho commented:
@mhegazy I just update the PR, please take a look when you have time.

mhegazy commented:
Sorry for the delay. I have manually merged it as there were merge conflicts, i have also marked the new declarations with the `/* @internal */` to avoid them showing in the .d.ts, moved the init code to a helper, and streamlined the serialization. take a look in case i messed something up.

mhegazy commented:
I wounder if this would be something interesting to editors; @basarat, and @jrieken do you see adding a feature in the editor to initialize tsconfig.json? if so we need to expose this in a public API/ tsserver command.

mhegazy commented:
@tinganho feel free to send your original changes for supporting comments in a separate PR.

jrieken commented:
@mhegazy - I like the feature but I think for editors it would be more useful if this would be exposed via ```tsserver```. As a nice bonus that way the server would also already know about the new project and wouldn't rely on another event.

mhegazy commented:
created #4466 to track the new feature.
